File: backend/scraper.py
import difflib
import random
import time

import requests
from bs4 import BeautifulSoup

USER_AGENTS = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4.1 Safari/605.1.15",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36 Edg/124.0.0.0",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 17_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4.1 Mobile/15E148 Safari/604.1",
    "Mozilla/5.0 (Linux; Android 14; Pixel 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Mobile Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36",
]

_last_fetch_times: dict[str, float] = {}


def _domain(url: str) -> str:
    from urllib.parse import urlparse
    return urlparse(url).netloc


def _respect_rate_limit(url: str) -> None:
    domain = _domain(url)
    last = _last_fetch_times.get(domain, 0)
    elapsed = time.time() - last
    delay = random.uniform(10, 30)
    if elapsed < delay:
        time.sleep(delay - elapsed)
    _last_fetch_times[domain] = time.time()


def _fetch_with_playwright(url: str, timeout: int = 30) -> str:
    from playwright.sync_api import sync_playwright
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page(user_agent=random.choice(USER_AGENTS))
        page.goto(url, timeout=timeout * 1000)
        content = page.content()
        browser.close()
    return content


def fetch_html(url: str, timeout: int = 30) -> str:
    _respect_rate_limit(url)
    headers = {"User-Agent": random.choice(USER_AGENTS)}
    resp = requests.get(url, headers=headers, timeout=timeout)
    resp.raise_for_status()
    html = resp.text

    soup = BeautifulSoup(html, "html.parser")
    body = soup.find("body")
    if body and len(body.get_text(strip=True)) < 50:
        html = _fetch_with_playwright(url, timeout)

    return html


def diff_html(old_html: str, new_html: str) -> str:
    if old_html == new_html:
        return ""
    old_lines = old_html.splitlines(keepends=True)
    new_lines = new_html.splitlines(keepends=True)
    diff = list(difflib.unified_diff(old_lines, new_lines, lineterm=""))
    return "".join(diff)

File: lib/scraper.ts
import axios from 'axios';
import * as cheerio from 'cheerio';
import { createTwoFilesPatch } from 'diff';

const USER_AGENTS = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4.1 Safari/605.1.15',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
];

const lastFetchTimes = new Map<string, number>();

function getDomain(url: string): string {
  return new URL(url).hostname;
}

async function respectRateLimit(url: string): Promise<void> {
  const domain = getDomain(url);
  const last = lastFetchTimes.get(domain) ?? 0;
  const elapsed = Date.now() - last;
  const delay = (10 + Math.random() * 20) * 1000; // 10–30s in ms
  if (elapsed < delay) {
    await new Promise((r) => setTimeout(r, delay - elapsed));
  }
  lastFetchTimes.set(domain, Date.now());
}

async function fetchWithPlaywright(url: string): Promise<string> {
  const { chromium } = await import('playwright');
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage({
    userAgent: USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)],
  });
  await page.goto(url, { timeout: 30_000 });
  const content = await page.content();
  await browser.close();
  return content;
}

async function fetchOnce(url: string): Promise<string> {
  // Vercel Serverless 跳过 Playwright（50MB 函数大小限制会炸）
  if (process.env.VERCEL) {
    const ua = USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)];
    const resp = await axios.get<string>(url, {
      headers: { 'User-Agent': ua },
      timeout: 30_000,
      responseType: 'text',
    });
    return resp.data;
  }

  const ua = USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)];
  const resp = await axios.get<string>(url, {
    headers: { 'User-Agent': ua },
    timeout: 30_000,
    responseType: 'text',
  });
  const html: string = resp.data;
  const $ = cheerio.load(html);
  const bodyText = $('body').text().trim();
  if (bodyText.length < 50) {
    return fetchWithPlaywright(url);
  }
  return html;
}

export async function fetchHtml(url: string, retries = 3): Promise<string> {
  await respectRateLimit(url);
  let lastError: unknown;
  for (let i = 0; i < retries; i++) {
    try {
      return await fetchOnce(url);
    } catch (err) {
      lastError = err;
      const wait = Math.min(4 * 2 ** i, 60) * 1000;
      await new Promise((r) => setTimeout(r, wait));
    }
  }
  throw lastError;
}

export function diffHtml(oldHtml: string, newHtml: string): string {
  if (oldHtml === newHtml) return '';
  return createTwoFilesPatch('old', 'new', oldHtml, newHtml, '', '', { context: 3 });
}
